bdb/rep.c

```diff
@@ -3729,6 +3732,85 @@ void bdb_set_seqnum(void *in_bdb_state)
 
 int gbl_online_recovery = 1;
 
+static pthread_mutex_t rep_mon_lk = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t rep_mon_cd = PTHREAD_COND_INITIALIZER;
+static pthread_t rep_mon_td;
+
+struct rep_mon {
+    pid_t tid;
+    char *host;
+    int starttime;
+    int type;
+};
+static hash_t *rep_mon_hash = NULL;
+
+int gbl_rep_mon_threshold = 0;
+
+static int process_rep_mon_hash(void *obj, void *arg)
+{
+    struct rep_mon *rm = (struct rep_mon *)obj;
+    int now = comdb2_time_epoch(), threshold = gbl_rep_mon_threshold; 
+    if (threshold > 0 && now - rm->starttime > threshold) {
+        fprintf(stderr, "Thread %u host %s type %d hung for %d seconds\n",
+            rm->tid, rm->host, rm->type, now - rm->starttime);
+        char pstack_cmd[128];
+        snprintf(pstack_cmd, sizeof(pstack_cmd), "pstack %d", (int)rm->tid);
+        system(pstack_cmd);
+        /* Linux pstacks tid, non-linux halts hash-for */
+#ifndef _LINUX_SOURCE
+        return 1;
+#endif
+    }
+    return 0;
+}
+
+static void *rep_mon(void *thd)
+{
+    pthread_mutex_lock(&rep_mon_lk);
+    if (rep_mon_hash) {
+        return NULL;
```

Comment:
I believe you forgot to unlock `rep_mon_lk` here. However, I think `rep_mon_hash` will never be NULL here, so it's better to replace this with an assert.

Reply:
Good catch!  The hash toggles to NULL if someone disables the tunable

Reply:
Ah ok, I thought it was not a dynamic tunable.